src/ProducerEventProcessor/include/ProducerEventProcessor/UploaderCaptureEventCollector.h

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef CAPTURE_EVENT_PROCESSOR_UPLOADER_CAPTURE_EVENT_COLLECTOR_H_
+#define CAPTURE_EVENT_PROCESSOR_UPLOADER_CAPTURE_EVENT_COLLECTOR_H_
+
+#include <absl/base/thread_annotations.h>
+#include <absl/synchronization/mutex.h>
+
+#include <memory>
+#include <vector>
+
+#include "CaptureFile/CaptureFileOutputStream.h"
+#include "CaptureUploader/UploadDataInterface.h"
+#include "ProducerEventProcessor/ClientCaptureEventCollector.h"
+#include "capture.pb.h"
+#include "services.grpc.pb.h"
+
+namespace orbit_producer_event_processor {
+
+// This class receives the ClientCaptureEvents emitted by a ProducerEventProcessor, buffers them
+// as a part of a well formated capture file, and enables a CaptureUploader to streaming upload the
+// buffered capture data.
+class UploaderCaptureEventCollector final : public ClientCaptureEventCollector,
+                                            public orbit_capture_uploader::UploadDataInterface {
+ public:
+  explicit UploaderCaptureEventCollector();
+  ~UploaderCaptureEventCollector() override;
+
+  // Covert the received capture event into raw bytes according to the format of capture file, and
+  // then buffer the converted data.
+  void AddEvent(orbit_grpc_protos::ClientCaptureEvent&& event) override;
+
+  // Functions needed by the CaptureUploader to upload data.
+  orbit_capture_uploader::DataReadiness GetDataReadiness() const override;
+  void RefreshUploadDataBuffer() override;
+  const std::vector<unsigned char>& GetUploadDataBuffer() const override {
+    return capture_data_to_upload_;
+  }
```

Comment:
Is there a reason for splitting this into 2 functions instead of having one (`TakeData`?) that returns the data by value?

Reply:
If there is no reason for having these functions separated then I would consider doing this (in a separate PR). But that's not super pressing.